### DISCUSSION TITLE
Game server with private sockets

### DISCUSSION THREAD

**Discussion by gus3inov** [miguelgrinberg/python-socketio#973] [Q&A] [answered]
Hi!
First of all, thanks for this library with a convenient api!
Want to consider the task, where we need send private sockets to users.

There is a game that has `players` and `enemy bots` that players have to kill. We only want to send players the sockets of the bots they are fighting, not all the bots in the game to save traffic on the client.
I implemented this solution via [socketio rooms](https://python-socketio.readthedocs.io/en/latest/server.html#rooms):
1. In `on_connect` of namespace I get the player's `sid` and store that `sid` in a redis map, where key is the player's username and value is the player's `sid`;
2. Next, in the background task where we create enemy-bots for the player, I am running the following algorithm: 
   1. getting the `sid` of the player who currently has few bots;
   2. create a enemy-bot and calling `sio_server.enter_room(sid=sid_player, room=bot_name)` . In this step we assign bot to player;
   3. When we want to change the bot information, we call `sio_server.emit(event_name, data, room=bot_name)` and the player should see the actual information about bot in interface, because player is in the room;
   4. If the bot has been killed, we call the `sio_server.leave_room(sid=sid_player, room=bot_name)` . In this step we detach the bot from the player;
   
The problem with this solution is that when the player reloads the page, the player's sid changes, then you need to get the rooms of the old sid and make `enter_room` with bots for the new player's sid, I think this is too complicated and not acanonical solution. 

Questions:
- Do you have any ideas on how to make it simpler and easier?
- How to get a list of rooms by player's sid?

P.S I will be grateful for any help in these matters!

**Comment by miguelgrinberg** [comment, marked as answer]
The `sid` is not great to be used as user identification, because as you well noticed, it changes when there is a reconnection (not only due to a refresh, but also due to network problems outside of the user's control). What you need to do is map the `sid` to your own user ID in the connect handler, then use your own user IDs when you map to your game entities. Then if you receive a reconnection you can update your user ID to the new `sid` value.

To get the rooms a client is in, use the [rooms() method](https://python-socketio.readthedocs.io/en/latest/api.html#socketio.Server.rooms).

**Comment by gus3inov** [comment]
Thanks for the answer!

Do I right get your take, I need to pass the user ID into the enter_room function instead of sid like `enter_room(sid=user_id)`?
Now in `on_connect` function I mapping `user_id: sid` and I don’t quite understand how my solution differs from the solution you suggested.

**Reply by miguelgrinberg** [reply]
No, that is not what I meant. What I mean is that you should not use `sid` values as identifiers for your users. You are doing that now, correct?

When you work with this package you have to use the `sid` values, but the state of your game should reference your own user identifiers, not the Socket.IO `sid`. If you store the room you assigned to a user in some sort of structure, let's say a dict with keys set to user IDs and values set to the room names, whenever a user reconnects you can look them up in your dict and immediately restore them to their rooms during the reconnection. I hope that makes it more clear.
